File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/basics/delegated_properties/java/ConfigurationService.java
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.delegated_properties.java;

public class ConfigurationService {

    private String configuration;

    public String getConfiguration() {
        // Java requires manual lazy initialization.
        if (configuration == null) {
            System.out.println("Loading configuration...");
            configuration = "application-config";
        }

        return configuration;
    }

    public void printConfiguration() {
        System.out.println(getConfiguration());
    }
}

File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/basics/delegated_properties/kotlin/ConfigurationService.kt
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.delegated_properties.kotlin

class ConfigurationService {

    // lazy is a delegated property.
    // The value is computed only when it is accessed for the first time.
    val configuration: String by lazy {
        println("Loading configuration...")
        "application-config"
    }

    fun printConfiguration() {
        println(configuration)
    }
}